jupyter_core/utils/__init__.py

```diff
@@ -163,18 +167,25 @@ def wrapped(*args, **kwargs):
     return wrapped
 
 
-async def ensure_async(obj: Union[Awaitable[Any], Any]) -> Any:
+# Matthias I believe this is technically incorrect,
```

Comment:
Was this meant to stay here as commentary?

Reply:
Removed. I'm still unsure about the correctness.

Reply:
I think we need it to be any awaitable since we use it for Futures too.

Reply:
Yeah, I don't think that's a big deal, but if it were used in a context where a async function return an awaitable on purpose then it's going then it may do the wrong things, 

I thing that 

`ensure_async(fun(val))` should likely be `ensure_async_func(func)(val)`, so you test the function, and not it's return value. I don't think it matters in how we are using it, but I think it's not the best abstraction.

Reply:
We could use `overload` to handle the different  cases

Reply:
But yeah, we should have changed the API to something better when we moved it to core

Reply:
Actually, no, this function is not meant to accept functions, only the result of calling a function, which could either be an awaitable or not.